### DISCUSSION TITLE
How i can configure clangd to use mingw-w64 includes

### DISCUSSION THREAD

**Discussion by ghost** [clangd/clangd#2234] [General]
when i install visual studio the clangd work with the vs includes (i dont have clang compiler), but when i switch to mingw g++ he give me an error because ha can't foud iostream or ostream.
please help, 'I'm using Neovim and clangd with mason'

**Comment by HighCommander4** [comment]
Please have a look at https://clangd.llvm.org/troubleshooting#cant-find-standard-library-headers-map-stdioh-etc. `--query-driver` tends to work in these cases.

**Comment by ghost** [comment]
i saw that and i fix it but now the problem exist just with 'float.h' it autocomplete it but give after that and error:
'''
    ~\Documents\C++\Projects\OpenGL\srcs\  1 
   └╴  main.cpp  1 
     └╴  In included file: 'float.h' file not found clang (pp_file_not_found) [4, 10]
'''

**Reply by HighCommander4** [reply]
Can you attach [clangd logs](https://clangd.llvm.org/troubleshooting.html#gathering-logs) for further diagnosis?

**Comment by ghost** [comment]
[START][2024-11-30 13:30:47] LSP logging initiated
[ERROR][2024-11-30 13:30:47] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:47.106] clangd version 19.1.0\r\nI[13:30:47.109] Features: windows\r\nI[13:30:47.109] PID: 4680\r\nI[13:30:47.109] Working directory: C:\\Users\\ucer\\Documents\\C++\\Projects\\OpenGL\r\nI[13:30:47.109] argv[0]: C:\\Program Files\\LLVM\\bin\\clangd.exe\r\nI[13:30:47.140] Starting LSP over stdin/stdout\r\nI[13:30:47.141] <-- initialize(1)\r\nI[13:30:47.149] --> reply:initialize(1) 7 ms\r\n"
[ERROR][2024-11-30 13:30:47] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:47.299] <-- initialized\r\nI[13:30:47.300] <-- textDocument/didOpen\r\nI[13:30:47.304] <-- textDocument/documentSymbol(2)\r\n"
[ERROR][2024-11-30 13:30:47] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:47.307] --> textDocument/publishDiagnostics\r\n"
[ERROR][2024-11-30 13:30:47] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	'I[13:30:47.340] Loaded compilation database from C:\\Users\\ucer\\Documents\\C++\\Projects\\OpenGL\\compile_commands.json\r\nI[13:30:47.342] ASTWorker building file C:\\Users\\ucer\\Documents\\C++\\Projects\\OpenGL\\srcs\\main.cpp version 0 with command \r\n[C:/Users/ucer/Documents/C++/Projects/OpenGL/build/x64]\r\n"C:\\\\mingw64\\\\bin\\\\g++.exe" --driver-mode=g++ -DGLEW_STATIC -IC:/Users/ucer/Documents/C++/Projects/OpenGL/./include -IC:/Users/ucer/Documents/C++/Projects/OpenGL/./srcs/imgui -g -std=gnu++17 -o CMakeFiles/OpenGL.dir/srcs/main.cpp.obj -c -D__SIZEOF_FLOAT128__=0 -D__GCC_MAX_ALIGN_T -D_LIBCPP_HAS_NO_MAX_ALIGN_T --target=x86_64-w64-mingw32 --gcc-toolchain=C:/mingw64 "-I C:/mingw64/include" "-I C:/mingw64/x86_64-w64-mingw32/include" "-resource-dir=C:\\\\Program Files\\\\LLVM\\\\lib\\\\clang\\\\19" -- "C:\\\\Users\\\\ucer\\\\Documents\\\\C++\\\\Projects\\\\OpenGL\\\\srcs\\\\main.cpp"\r\nI[13:30:47.343] --> window/workDoneProgress/create(0)\r\nI[13:30:47.343] Enqueueing 8 commands for indexing\r\n'
[ERROR][2024-11-30 13:30:47] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:47.348] <-- textDocument/semanticTokens/full(3)\r\n"
[ERROR][2024-11-30 13:30:47] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:47.365] <-- reply(0)\r\nI[13:30:47.365] --> $/progress\r\nI[13:30:47.365] --> $/progress\r\n"
[ERROR][2024-11-30 13:30:49] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:49.574] Built preamble of size 4154792 for file C:\\Users\\ucer\\Documents\\C++\\Projects\\OpenGL\\srcs\\main.cpp version 0 in 2.23 seconds\r\n"
[ERROR][2024-11-30 13:30:49] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:49.576] --> workspace/semanticTokens/refresh(1)\r\n"
[ERROR][2024-11-30 13:30:49] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:49.576] Indexing c++17 standard library in the context of C:\\Users\\ucer\\Documents\\C++\\Projects\\OpenGL\\srcs\\main.cpp\r\nI[13:30:49.578] <-- $/cancelRequest\r\nI[13:30:49.578] <-- textDocument/semanticTokens/full(4)\r\nI[13:30:49.579] <-- reply(1)\r\n"
[ERROR][2024-11-30 13:30:49] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:49.617] --> textDocument/publishDiagnostics\r\n"
[ERROR][2024-11-30 13:30:49] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:49.618] --> reply:textDocument/documentSymbol(2) 2313 ms\r\nI[13:30:49.618] --> reply:textDocument/semanticTokens/full(3) 2269 ms, error: Task was cancelled.\r\nI[13:30:49.618] --> reply:textDocument/semanticTokens/full(4) 39 ms\r\n"
[ERROR][2024-11-30 13:30:51] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:51.482] <-- textDocument/documentSymbol(5)\r\n"
[ERROR][2024-11-30 13:30:51] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:51.482] --> reply:textDocument/documentSymbol(5) 0 ms\r\n"
[ERROR][2024-11-30 13:30:56] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:56.638] --> $/progress\r\nI[13:30:56.638] --> $/progress\r\n"
[ERROR][2024-11-30 13:30:57] .../vim/lsp/rpc.lua:770	"rpc"	"C:\\Program Files\\LLVM\\bin\\clangd.EXE"	"stderr"	"I[13:30:57.320] Indexed c++17 standard library (incomplete due to errors): 11711 symbols, 574 filtered\r\n"

**Reply by HighCommander4** [reply]
You're not using `--query-driver` in this run.

Also, please add `--log=verbose` when taking any further logs.

**Comment by ghost** [comment]
no problem was solved now,
i just need to configure .clangd to:
```
CompileFlags:
  Add:
    - "--gcc-toolchain=C:/mingw64"
    - -D__SIZEOF_FLOAT128__=0
    - --target=x86_64-w64-mingw32
    - -D__GCC_MAX_ALIGN_T
    - -D_LIBCPP_HAS_NO_MAX_ALIGN_T
    - "-isystem"
    - "C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/include/c++"
    - "-isystem"
    - "C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/include/c++/x86_64-w64-mingw32"
    - "-isystem"
    - "C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/include/c++/backward"
    - "-isystem"
    - "C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/include"
    - "-isystem"
    - "C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/include-fixed"
    - "-isystem"
    - "C:/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/8.1.0/../../../../x86_64-w64-mingw32/include"
```
